prow/config/config.go

```diff
@@ -1239,43 +1258,141 @@ type Branding struct {
 type RerunAuthConfigs map[string]prowapi.RerunAuthConfig
 
-// GetRerunAuthConfig returns the appropriate RerunAuthConfig based on the provided Refs.
-func (rac RerunAuthConfigs) GetRerunAuthConfig(refs *prowapi.Refs) prowapi.RerunAuthConfig {
-	if refs == nil || refs.Org == "" {
-		return rac["*"]
+type DefaultRerunAuthConfigEntry struct {
+	// Matching/filtering fields. All filters must match for an entry to match.
+
+	// OrgRepo matches against the "org" or "org/repo" that the presubmit or postsubmit
+	// is associated with. If the job is a periodic, extra_refs[0] is used. If the
+	// job is a periodic without extra_refs, the empty string will be used.
+	// If this field is omitted all jobs will match.
+	OrgRepo string `json:"repo,omitempty"`
+	// Cluster matches against the cluster alias of the build cluster that the
+	// ProwJob is configured to run on. Recall that ProwJobs default to running on
+	// the "default" build cluster if they omit the "cluster" field in config.
+	Cluster string `json:"cluster,omitempty"`
+
+	// Config is the RerunAuthConfig to apply if the filter fields all match the
+	// ProwJob. Note that when multiple entries match a ProwJob the entry with the
+	// highest specification is used.
+	Config *prowapi.RerunAuthConfig `json:"rerun_auth_configs,omitempty"`
+}
+
+func (d *Deck) GetRerunAuthConfig(refs *prowapi.Refs, cluster string) *prowapi.RerunAuthConfig {
+	var config *prowapi.RerunAuthConfig
+
+	var orgRepo string
+	if refs != nil {
+		orgRepo = refs.OrgRepoString()
 	}
 
-	if rerun, exists := rac[fmt.Sprintf("%s/%s", refs.Org, refs.Repo)]; exists {
-		return rerun
+	for _, drac := range d.DefaultRerunAuthConfigs {
+		if drac.matches(orgRepo, cluster) {
+			config = drac.Config
+		}
 	}
 
-	if rerun, exists := rac[refs.Org]; exists {
-		return rerun
+	return config
+}
+
+// defaultRerunAuthMapToSlice converts the old format
+// (map[string]*prowapi.RerunAuthConfig) to the new format
+// ([]*DefaultRerunAuthConfigEntry) or DefaultRerunAuthConfigs.
+func defaultRerunAuthMapToSlice(m map[string]prowapi.RerunAuthConfig) ([]*DefaultRerunAuthConfigEntry, error) {
+	mLength := len(m)
+	var entries []*DefaultRerunAuthConfigEntry
+	add := func(repo string, rac prowapi.RerunAuthConfig) {
+		entries = append(entries, &DefaultRerunAuthConfigEntry{
+			OrgRepo: repo,
+			Cluster: "",
+			Config:  &rac,
+		})
+	}
+	// Safe to delete from map because RerunAuthConfigs is deprecated.
+	remove := func(repo string) {
+		delete(m, repo)
+	}
+
+	// Ensure "" comes first...
+	if rac, ok := m[""]; ok {
+		add("", rac)
+		remove("")
+	}
+	// Ensure "*" comes first...
+	if rac, ok := m["*"]; ok {
+		add("*", rac)
+		remove("*")
+	}
+	// then orgs...
+	for key, rac := range m {
+		if strings.Contains(key, "/") {
+			continue
+		}
+		add(key, rac)
+		remove(key)
+	}
+	// then repos.
+	for key, rac := range m {
+		add(key, rac)
+	}
+
+	var err error
+	if mLength != len(entries) {
```

Comment:
This validation will always fail if there is an entry for `*` or empty string in the map, as you delete those from the map prior to iterating it. I would also recommend some tests, so this kind of logic mistake gets noticed.

Also this code is needlessly complicated, please just `return fmt.Errorf(....)`

Reply:
mLength represents the length of the map before any elements are removed from it. Here we are verifying that the new slice that we created, `entries`, has the same amount of elements at the end before elements are removed from the map. Thus, removing from the map would not affect the condition here.
 
There are test cases that contain `*` or empty string in the map. View `TestRerunAuthConfigsGetRerunAuthConfig` for a reference. They currently work as expected.

The function now just returns the error directly.

Reply:
Oh you are right, I missed the early initialization of `mLength`